File: frontend/src/types/commerce.ts
import { Account, Contact } from './crm';

export interface Product {
  id: string;
  name: string;
  sku: string;
  price: number;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface QuotationItem {
  id: string;
  quotation_id: string;
  product_id: string;
  quantity: number;
  unit_price: number;
  discount: number;
  subtotal: number;
  total: number;
  product?: Product;
}

export type QuotationStatus =
  | 'draft'
  | 'pending_approval'
  | 'approved'
  | 'rejected'
  | 'sent'
  | 'expired'
  | 'accepted';

export interface Quotation {
  id: string;
  account_id: string;
  contact_id: string | null;
  quote_number: string;
  status: QuotationStatus;
  subtotal: number;
  discount_type: 'percentage' | 'fixed';
  discount_value: number;
  discount_amount: number;
  tax_rate: number;
  tax_amount: number;
  total_amount: number;
  valid_until: string | null;
  terms_conditions: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  items: QuotationItem[];
  account?: Account;
  contact?: Contact | null;
}

export interface OrderItem {
  id: string;
  order_id: string;
  product_id: string | null;
  quantity: number;
  unit_price: number;
  subtotal: number;
  total: number;
  product?: Product;
}

export type OrderStatus =
  | 'pending'
  | 'confirmed'
  | 'processing'
  | 'shipped'
  | 'delivered'
  | 'cancelled';

export type PaymentStatus = 'unpaid' | 'partially_paid' | 'paid' | 'refunded';

export interface Order {
  id: string;
  quotation_id: string | null;
  account_id: string;
  contact_id: string | null;
  order_number: string;
  status: OrderStatus;
  payment_status: PaymentStatus;
  payment_method: string | null;
  shipping_address: string | null;
  billing_address: string | null;
  subtotal: number;
  tax_amount: number;
  total_amount: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
  items: OrderItem[];
  account?: Account;
  contact?: Contact | null;
  quotation?: Quotation | null;
}

export interface QuotationItemPayload {
  product_id: string;
  quantity: number;
  unit_price: number;
  discount: number;
}

export interface QuotationCreatePayload {
  account_id: string;
  contact_id?: string | null;
  discount_type: 'percentage' | 'fixed';
  discount_value: number;
  tax_rate: number;
  valid_until?: string | null;
  terms_conditions?: string | null;
  notes?: string | null;
  items: QuotationItemPayload[];
}

export interface OrderCreatePayload {
  account_id: string;
  contact_id?: string | null;
  shipping_address?: string | null;
  billing_address?: string | null;
  payment_method?: string | null;
  notes?: string | null;
  quotation_id?: string | null;
  items?: QuotationItemPayload[];
}

File: backend/app/models/commerce.py
import uuid
from datetime import datetime
from typing import List, Optional
from sqlalchemy import String, Text, Numeric, Integer, DateTime, ForeignKey, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import BaseModel


class Product(BaseModel):
    __tablename__ = "products"

    name: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    sku: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    price: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)


class Quotation(BaseModel):
    __tablename__ = "quotations"

    account_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("accounts.id", ondelete="CASCADE"), nullable=False)
    contact_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("contacts.id", ondelete="SET NULL"), nullable=True)
    quote_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(50), default="draft")  # draft, pending_approval, approved, rejected, sent, expired, accepted
    
    subtotal: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    discount_type: Mapped[str] = mapped_column(String(20), default="percentage")  # percentage, fixed
    discount_value: Mapped[float] = mapped_column(Numeric(15, 2), default=0.0)
    discount_amount: Mapped[float] = mapped_column(Numeric(15, 2), default=0.0)
    tax_rate: Mapped[float] = mapped_column(Numeric(5, 2), default=18.0)
    tax_amount: Mapped[float] = mapped_column(Numeric(15, 2), default=0.0)
    total_amount: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    
    valid_until: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    terms_conditions: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Relationships
    account: Mapped["Account"] = relationship("Account")
    contact: Mapped[Optional["Contact"]] = relationship("Contact")
    items: Mapped[List["QuotationItem"]] = relationship(
        "QuotationItem", back_populates="quotation", cascade="all, delete-orphan"
    )


class QuotationItem(BaseModel):
    __tablename__ = "quotation_items"

    quotation_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("quotations.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("products.id", ondelete="SET NULL"), nullable=True)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    unit_price: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    discount: Mapped[float] = mapped_column(Numeric(15, 2), default=0.0)  # item level discount
    subtotal: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    total: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)

    # Relationships
    quotation: Mapped["Quotation"] = relationship("Quotation", back_populates="items")
    product: Mapped[Optional[Product]] = relationship("Product")


class Order(BaseModel):
    __tablename__ = "orders"

    quotation_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("quotations.id", ondelete="SET NULL"), nullable=True)
    account_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("accounts.id", ondelete="CASCADE"), nullable=False)
    contact_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("contacts.id", ondelete="SET NULL"), nullable=True)
    order_number: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    status: Mapped[str] = mapped_column(String(50), default="pending")  # pending, confirmed, processing, shipped, delivered, cancelled
    payment_status: Mapped[str] = mapped_column(String(50), default="unpaid")  # unpaid, partially_paid, paid, refunded
    payment_method: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    
    shipping_address: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    billing_address: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    subtotal: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    tax_amount: Mapped[float] = mapped_column(Numeric(15, 2), default=0.0)
    total_amount: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    notes: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Relationships
    account: Mapped["Account"] = relationship("Account")
    contact: Mapped[Optional["Contact"]] = relationship("Contact")
    quotation: Mapped[Optional[Quotation]] = relationship("Quotation")
    items: Mapped[List["OrderItem"]] = relationship(
        "OrderItem", back_populates="order", cascade="all, delete-orphan"
    )


class OrderItem(BaseModel):
    __tablename__ = "order_items"

    order_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("orders.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("products.id", ondelete="SET NULL"), nullable=True)
    quantity: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    unit_price: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    subtotal: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)
    total: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False, default=0.0)

    # Relationships
    order: Mapped["Order"] = relationship("Order", back_populates="items")
    product: Mapped[Optional[Product]] = relationship("Product")
